thrust/testing/complex.cu

```diff
@@ -7,70 +8,231 @@
+      move_assign = thrust::complex<T>{real, imag};
+      ASSERT_EQUAL(expected.real(), move_assign.real());
+      ASSERT_EQUAL(expected.imag(), move_assign.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_lvalue_T{};
+      const thrust::complex<T> expected{real};
+      const T to_be_copied = real;
+      assign_from_lvalue_T = to_be_copied;
+      ASSERT_EQUAL(expected.real(), assign_from_lvalue_T.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_lvalue_T.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_rvalue_T{};
+      const thrust::complex<T> expected{real};
+      assign_from_rvalue_T = T(real);
+      ASSERT_EQUAL(expected.real(), assign_from_rvalue_T.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_rvalue_T.imag());
+    }
+
+    {
+      const std::complex<T> expected(real, imag);
+      const thrust::complex<T> copy_from_std(expected);
+      ASSERT_EQUAL(expected.real(), copy_from_std.real());
+      ASSERT_EQUAL(expected.imag(), copy_from_std.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_lvalue_std{};
+      const std::complex<T> expected(real, imag);
+      assign_from_lvalue_std = expected;
+      ASSERT_EQUAL(expected.real(), assign_from_lvalue_std.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_lvalue_std.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_rvalue_std{};
+      assign_from_rvalue_std = std::complex<T>(real, imag);
+      ASSERT_EQUAL(real, assign_from_rvalue_std.real());
+      ASSERT_EQUAL(imag, assign_from_rvalue_std.imag());
+    }
   }
 };
-SimpleUnitTest<TestComplexConstructors, FloatingPointTypes> TestComplexConstructorsInstance;
+SimpleUnitTest<TestComplexConstructionAndAssignment, FloatingPointTypes>
+  TestComplexConstructionAndAssignmentInstance;
 
+template <typename T>
+struct TestComplexConstructionAndAssignmentWithPromoting
+{
+  void operator()(void)
+  {
+    using T0 = T;
+    typedef other_floating_point_type_t<T0>  T1;
+
+    thrust::host_vector<T1> data = unittest::random_samples<T1>(2);
+
+    const T0 real_T0 = data[0];
+    const T0 imag_T0 = data[1];
+
+    const T1 real_T1 = static_cast<T1>(real_T0);
+    const T1 imag_T1 = static_cast<T1>(imag_T0);
+
+    {
+      const thrust::complex<T0> construct_from_real_and_imag(real_T1, imag_T1);
+      ASSERT_ALMOST_EQUAL(real_T0, construct_from_real_and_imag.real());
+      ASSERT_ALMOST_EQUAL(imag_T0, construct_from_real_and_imag.imag());
+    }
+
+    {
+      const thrust::complex<T0> construct_from_real(real_T1);
+      ASSERT_ALMOST_EQUAL(real_T1, construct_from_real.real());
+      ASSERT_EQUAL(T(), construct_from_real.imag());
+    }
+
+    {
+      const thrust::complex<T1> expected(real_T1, imag_T1);
+      thrust::complex<T0> construct_from_copy(expected);
+      ASSERT_ALMOST_EQUAL(expected.real(), construct_from_copy.real());
+      ASSERT_ALMOST_EQUAL(expected.imag(), construct_from_copy.imag());
```

Comment:
I believe we should be clear that we always want to compare against `T0`
```suggestion
      ASSERT_ALMOST_EQUAL(real_T0, construct_from_copy.real());
      ASSERT_ALMOST_EQUAL(imag_T0, construct_from_copy.imag());
```

Applies below